_posts/2015-08-06-machinist.md

```diff
@@ -0,0 +1,434 @@
+
+Enter machinist
+---------------
+
+Machinist is based on a set of macros that were introduced in [Spire](https://github.com/non/spire) to remove the performance penalties associated with generic math implementations. These macros were based on an even earlier approach which used a compiler plugin.
+
+The basic approach has not changed: at compile-time we can detect situations where we build an object just to assemble a method call with the arguments to its constructor. In these cases we rewrite the tree, removing the object allocation and making the method call directly. Machinist's documentation goes to some trouble to explain it, but basically, we want to be able to write code like `Test3.gen` but have it interpreted as `Test1.gen`. That is literally the entire purpose of machinist.
+
+Here's a construction that works for this example:
+
+```scala
+object Test2 {
+  implicit class DivOps[A](lhs: A)(implicit ev: Div[A]) {
+    def /(rhs: A): A = macro machinist.DefaultOps.binop[A, A]
+  }
+}
+
+class Test2 {
+  import Test2.DivOps
+  def gen[A: Div](x: A, y: A): A = x / y
+  def test: String = gen("foo", "bar")
+}
+```
+
+We use the `machinist.DefaultOps` object to provide an instance of the `binop` macros, which will rewrite `DivOps(x)(ev).$div(y)` into `ev.div(x, y)`.
+
+Here's what we end up with in bytecode:
+
+```
+// cost.Test2.gen(A, A, Div[A]): A
+0: aload_3
+1: aload_1
+2: aload_2
+3: invokeinterface #26,  3           // InterfaceMethod cost/Div.div:(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
+8: areturn
+```
+
+As you can see, the sourcecode for `Test2.gen` is identical to `Test3.gen`, and the bytecode for `Test2.gen` is identical to that of `Test1.gen`. Success!
+
+Caveats
+-------
+
+There are a few caveats that are worth mentioning:
+
+### Managing compilation units
+
+The issue that sparked this article used the operator `+/+`. Machinist claims to be able to support any symbolic operator. Why didn't we use that operator here?
+
+The answer has to do with how Scala macros work. Scala requires that macros be defined in a separate "compilation unit" from the one they are invoked in. This makes it very awkward to create a code snippet that both defines and uses a macro. In this case, it means that we can't extend `machinist.Ops` to define new symbolic operators in the same file that demonstrates their use. This is why we used `/` (which maps to `div` and is a "default operator").
+
+### Use outside of generic methods
+
+Now that we've demonstrated the cost that implicit conversions to value classes impose, you might imagine wanting to perform this transformation on *all* your implicit conversions.
+
+Unfortunately, Machinist is not sufficiently general to support this. Right now its macros support a number of different "shapes" but assume generic methods with type parameters. It might be possible to write macros which inline the method body of a concrete implicit class, but that's outside the scope of the project. 
```

Comment:
This restriction, as written, seems to exclude `div`: it is not a "generic method with type parameters".  Can you say exactly what the restrictions are here?